### PR TITLE
Refactor, add unittests and fix line-rate

### PR DESCRIPTION
**Pull request by skoppe**
The line-rate was wrong, which prompted a refactor and adding unittests

### CODE REVIEW

**Comment by nordlow** [comment]
Nitpick: template restrictions? :)

File: src/cobertura.d [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+struct LineCoverage {
+	size_t index;
+	size_t count;
+}
+
+struct FileCoverage {
+	string path;
+	LineCoverage[] lines;
+	size_t covered;
```
Comment (nordlow) [inline src/cobertura.d]: I would prefer a more descriptive naming containing the term "count".

File: src/cobertura.d [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+auto generateXmlFile(FileCoverage[] files) {
+	import core.stdc.time: time;
+	import std.format: format;
+	import std.path: buildPath;
+
+	import std.algorithm.iteration: splitter, map, sum;
+
+	size_t linesValid = files.map!(file => file.lines.length).sum();
+	size_t linesCovered = files.map!(file => file.covered).sum();
```
Comment (nordlow) [inline src/cobertura.d]: Could these two iterations be joined? I think I've asked this question in another context already. Don't remember the answer.

File: src/cobertura.d (hunk 1 above)
Comment (skoppe) [reply src/cobertura.d]: `linesCovered`?

File: src/cobertura.d (hunk 2 above)
Comment (skoppe) [reply src/cobertura.d]: `fold` can aggregate multiple things in one iteration.

`files.fold!((acc,f) => acc+f.lines.length, (acc,t) => acc+f.covered)(0,0);` would return a tuple with both things summed.

but it is ugly as hell, and without destructuring using it isn't pretty either.

**Comment by skoppe** [comment]
> Nitpick: template restrictions? :)

Where?

File: src/cobertura.d (hunk 1 above)
Comment (nordlow) [reply src/cobertura.d]: I'd prefer `lineCount` or `coveredLinesCount` or or `coverageLinesCount` or something like that. The term "coverage" is already in the struct name.

Btw, what's the difference between `lines.length` and `covered`?

File: src/cobertura.d [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+	LineCoverage[] lines;
+	size_t covered;
+}
+
+FileCoverage parseLstFile(string path) {
+	import std.stdio : File;
+	return parseLstFile(File(path).byLine, path);
+}
+
+template tryTo(T) {
```
Comment (nordlow) [inline src/cobertura.d:23]: Tempate restriction.

File: src/cobertura.d [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+		import std.conv : to;
+		try {
+			return nullable(r.to!T);
+		} catch (Exception e) {
+			return Nullable!T.init;
+		}
+	}
+}
+
+auto filterNullable(Range)(Range r) {
```
Comment (nordlow) [inline src/cobertura.d]: Tempate restriction.

File: src/cobertura.d [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+	import std.algorithm : filter, map;
+	return r.filter!(a => !a.isNull).map!(a => a.get);
+}
+
+private template andThen(alias fun) {
```
Comment (nordlow) [inline src/cobertura.d:34]: Tempate restriction.

File: src/cobertura.d [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        } else {
+            alias Result = Nullable!RT;
+            if (t.isNull)
+                return Result.init;
+            return Result(fun(t.get));
+        }
+    }
+}
+
+FileCoverage parseLstFile(RoR)(RoR r, string path) if (isInputRange!RoR &&
```
Comment (nordlow) [inline src/cobertura.d]: Tempate restriction.

File: src/cobertura.d (hunk 2 above)
Comment (nordlow) [reply src/cobertura.d]: Ok, fair enough.

File: src/cobertura.d (hunk 3 above)
Comment (skoppe) [reply src/cobertura.d:23]: What would the restriction be? I mean, it is just a passthrough function. It works for all T and R as long as `r.to!T` work. I could put that in the constraint, but what would the benefit be?

File: src/cobertura.d (hunk 4 above)
Comment (skoppe) [reply src/cobertura.d]: Yep. `ElementType!isNullable`. Knew it from the start but was lazy. Thanks.

File: src/cobertura.d (hunk 5 above)
Comment (skoppe) [reply src/cobertura.d:34]: I can restrict the fun, but what would the benefit be?

The error the compiler generates is pretty ok: https://run.dlang.io/is/4eVvQ4

File: src/cobertura.d (hunk 6 above)
Comment (skoppe) [reply src/cobertura.d]: There are 2 no? Has to be an inputRange with isSomeString elementtype.

File: src/cobertura.d (hunk 5 above)
Comment (nordlow) [reply src/cobertura.d:34]: I was thinking of

`if (isCallable!fun)`

or/and potentially some restriction that checks that `fun` is callable as `fun(T.init)`.

Yes, I know they seem overkill. But I'm thinking that in a not so distant future we could have some-clever intellisense that could be sped up by those restrictions. ;)

File: src/cobertura.d (hunk 6 above)
Comment (nordlow) [reply src/cobertura.d]: Correct.

Comment (nordlow) [reply src/cobertura.d]: I wish there was a way to "bottom-up" deduce the template restriction from how `r` is used in the template definition.

File: src/cobertura.d (hunk 1 above)
Comment (skoppe) [reply src/cobertura.d]: `lines` represents the lines that actually end up in the object file. Not all lines do, e.g. comments, static ifs, etc.

Each line has a counter that keeps track how many times it got executed.

`covered` represents the amount of lines that have non-zero counters.

File: src/cobertura.d (hunk 6 above)
Comment (nordlow) [reply src/cobertura.d]: Perhaps `Lines` is a better naming instead of RoR`.

File: src/cobertura.d (hunk 3 above)
Comment (nordlow) [reply src/cobertura.d:23]: Ahh, you're right. `std.conv.to` doesn't have any template restrictions. My mistake.

File: src/cobertura.d (hunk 1 above)
Comment (nordlow) [reply src/cobertura.d]: I see. Thanks.

File: src/cobertura.d (hunk 5 above)
Comment (skoppe) [reply src/cobertura.d:34]: Doesn't work, probably because `fun` is a templated lambda.

File: src/cobertura.d [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+			res ~= "\t\t\t\t\t\t<line number=\"%s\" hits=\"%s\"/>\n".format(line.index+1, line.count);
+		}
+		res ~= "\t\t\t\t\t</lines>\n\t\t\t\t</class>\n\t\t\t</classes>\n\t\t</package>\n\t";
+	}
+	res ~= "</packages>\n</coverage>";
+	return res;
+}
+
+unittest {
+	import unit_threaded;
```
Comment (burner) [inline src/cobertura.d:130]: pulling in this package for one call to should is overkill IMO

File: dub.sdl [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+configuration "default" {
+}
+
+configuration "unittest" {
+    dependency "unit-threaded" version="*"
+    mainSourceFile "bin/ut.d"
+    excludedSourceFiles "src/app.d"
+    preBuildCommands "$DUB run --compiler=$$DC unit-threaded -c gen_ut_main -- -f bin/ut.d -d $DUB"
+}
```
Comment (burner) [inline dub.sdl:17]: what is wrong with the json version?

File: src/cobertura.d [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+			return nullable(r.to!T);
+		} catch (Exception e) {
+			return Nullable!T.init;
+		}
+	}
+}
+
+enum isNullable(T) = is(T : Nullable!P, P);
+
+auto filterNullable(Range)(Range r) if (isNullable!(ElementType!Range)) {
```
Comment (burner) [inline src/cobertura.d]: the name is misleading. I would say it is getIfNotNull?

But as it is called once inline filter!(i => !i.isNull()).map!(i => i.get()) should have worked as well

File: src/cobertura.d [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+	import std.algorithm : filter, map;
+	return r.filter!(a => !a.isNull).map!(a => a.get);
+}
+
+private template andThen(alias fun) {
+    auto andThen(T)(Nullable!T t) {
+        alias RT = typeof(fun(T.init));
+        static if (is(RT == void)) {
+            if (!t.isNull)
```
Comment (burner) [inline src/cobertura.d]: this is a nitpick but, if statements should always be followed by block statements. Single statement version will always bite somebody down the road

Comment (skoppe) [reply src/cobertura.d]: Sure.

File: src/cobertura.d (hunk 9 above)
Comment (skoppe) [reply src/cobertura.d]: You are right. It filters *non-Nullable*, not Nullable. And then it unpacks them. Haskell probably has a name for it.

File: dub.sdl (hunk 8 above)
Comment (skoppe) [reply dub.sdl:17]: It is so ugly.

File: src/cobertura.d (hunk 7 above)
Comment (skoppe) [reply src/cobertura.d:130]: There are two calls actually, but you have a point.

That said, I never leave the house without unit_threaded

File: src/cobertura.d (hunk 9 above)
Comment (burner) [reply src/cobertura.d]: especially, as you could run the old thing by typing dmd -run source/app.d

Comment (skoppe) [reply src/cobertura.d]: Haskell calls it [`catMaybes`](https://hackage.haskell.org/package/base-4.7.0.2/docs/Data-Maybe.html#v:catMaybes)

Scala calls it `flatten`.

I'll go with `flatten`

File: dub.sdl (hunk 8 above)
Comment (burner) [reply dub.sdl:17]: not as much as the sdl version ;-)

File: src/cobertura.d (hunk 9 above)
Comment (burner) [reply src/cobertura.d]: my suggestion would be to remove it, as its called in one place

Comment (skoppe) [reply src/cobertura.d]: Sure.

File: dub.sdl (hunk 8 above)
Comment (skoppe) [reply dub.sdl:17]: Well, besides the aesthetics, we mostly use `dub.sdl`'s. Unless you insist I'll leave it at that.

**Comment by burner** [comment]
unit threaded is still used, my comment about being able to run "dmd -run source/app.d" seemed to be to have been deleted

**Comment by skoppe** [comment]
> my comment about being able to run "dmd -run source/app.d" seemed to be to have been deleted

I saw it appear somewhere else. But why is being able to run `dmd -run src/app.d` so important?

`dmd -run <srcfile>` only works on single source applications. Now that I have split things into two it doesn't work.

This does though: `dmd -I=src/ -i -run src/app.d` or just `cd src; dmd -i -run app.d`.

> unit threaded is still used

It has the added benefit of printing nice expected messages.

**Comment by burner** [comment]
why was it split into two anyway? Pardon me, I think today I'm grumpy cat.

Anyway, moving on.

**Comment by skoppe** [comment]
> why was it split into two anyway?

Because of unit threaded 😄  But it is good hygiene regardless.

> Pardon me, I think today I'm grumpy cat.

No problem.

**Pull request merged**